ocaml/typing/solver_intf.ml

```diff
@@ -0,0 +1,298 @@
+    val constrain_upper : 'a C.obj -> ('a, 'e * allowed) mode -> 'a
+
+    (** Create a new mode above the given mode *)
+    val newvar_above :
+      'a C.obj -> ('a, allowed * 'r_) mode -> ('a, 'l * 'r) mode * bool
+
+    (** Create a new mode below the given mode *)
+    val newvar_below :
+      'a C.obj -> ('a, 'l_ * allowed) mode -> ('a, 'l * 'r) mode * bool
+
+    (** Give the least upper bound of the list of modes *)
+    val join : 'a C.obj -> ('a, allowed * 'r) mode list -> ('a, left_only) mode
+
+    (** Give the greatest lower bound of the list of modes *)
+    val meet : 'a C.obj -> ('a, 'l * allowed) mode list -> ('a, right_only) mode
+
+    (** Check if the mode has been constrained into a constant *)
+    val check_const : 'a C.obj -> ('a, 'd0 * 'd1) mode -> 'a option
+
+    (** Print the given mode.
+    [verbose] decides whether mode variable identifiers are printed instead of a
+    question mark. *)
+    val print :
+      ?verbose:bool ->
+      'a C.obj ->
+      Format.formatter ->
+      ('a, 'l * 'r) mode ->
+      unit
+
+    val print_raw :
+      ?verbose:bool ->
+      'a C.obj ->
+      Format.formatter ->
+      ('a, 'l * 'r) mode ->
+      unit
+  end
+
+  (** Extending [Solver_mono] to support dualized objects. This is needed
+   because the morphisms between uniqueness and linearity can be anti-tone. *)
+
+  module Solver_polarized (C : Lattices_mono) : sig
+    module S : module type of Solver_mono (C)
+
+    (* First construct a new category based on the original category C. The
+       objects are those from the C, and those from C but flipped lattice
+       structure. The morphisms are the obvious four copies of the original
+       morhpisms. *)
+
+    type 'a obj =
+      | Positive : 'a C.obj -> ('a * positive) obj
+          (** The original lattice of obj *)
+      | Negative : 'a C.obj -> ('a * negative) obj
+          (** the dual lattice of obj *)
+
+    (* ['a] and ['b] are source and destination objects;
+       ['d] and ['e] are source and desitnation adjoint status *)
+    type ('a, 'd, 'b, 'e) morph =
+      | Pos_Pos :
+          ('a, 'b, 'd) C.morph
+          -> ('a * positive, 'd pos, 'b * positive, 'd pos) morph
+      | Pos_Neg :
+          ('a, 'b, 'd) C.morph
+          -> ('a * positive, 'd pos, 'b * negative, 'd neg) morph
+      | Neg_Pos :
+          ('a, 'b, 'd) C.morph
+          -> ('a * negative, 'd neg, 'b * positive, 'd pos) morph
+      | Neg_Neg :
+          ('a, 'b, 'd) C.morph
+          -> ('a * negative, 'd neg, 'b * negative, 'd neg) morph
+
+    (* [id] and [compose] not used; just for fun *)
+    val id : 'a obj -> ('a, 'l * 'r, 'a, 'l * 'r) morph
+
+    val compose :
+      'c obj ->
+      ('b, 'bl * 'br, 'c, 'cl * 'cr) morph ->
+      ('a, 'al * 'ar, 'b, 'bl * 'br) morph ->
+      ('a, 'al * 'ar, 'c, 'cl * 'cr) morph
+
+    type ('a, 'd) mode =
+      | Positive : ('a, 'd) S.mode -> ('a * positive, 'd pos) mode
```

Comment:
As far as I can tell, these constructors don't really serve a purpose. That is, every pattern match on this type already knows what constructor it will find before the match. I see that they do stuff to types. But I wonder if there's a way to encode the idea without using constructors, which take up space at runtime (and also are confusing if they don't really matter).

Reply:
All function in this module (`Solver_polarized`) are very thin wrappers so hopefully inlined. We can add explicit annotation to ensure that.

I feel that these functions being thin wrappers are correlated to the comment you made.

Reply:
I agree that the wrappers are likely inlined. But that doesn't necessarily mean that these constructors never appear at runtime, especially when the constructors are applied inside of tuples, for example.

Reply:
I think we could probably use:
```ocaml
type ('a, 'd) mode = Mode : ('a, 'd) S.mode -> ('a * _, 'd pos) mode
  [@@unboxed]
```
to avoid the runtime cost of the constructors. Note that this does force us into the design where you need to pass in the associated `obj` whenever you try to operate on one of these, since it will be the only way to learn whether a mode is positive or negative. That's our current design anyway though, and I also prefer it for other reasons.

Reply:
I don't think that would work, because it ties the `'d` of the inner mode (to be used with the monotone solver) to the `'d` of the outer mode (to be used with the polarized solver). That correlation is false for negative mode axes. I've spent an embarrassing amount of time thinking about this and how to do it better. I might have a way forward, but I've exhausted my time budget here.

Reply:
This is resolved, now that we have merged Richard's PR.